File: dohproxy/dnsutil.py
"""Tiny helpers to make DNS queries human-readable for logging."""

from __future__ import annotations

import struct

# Common QTYPE numbers -> names (anything else is shown as TYPE<n>).
_QTYPES = {
    1: "A", 2: "NS", 5: "CNAME", 6: "SOA", 12: "PTR", 15: "MX", 16: "TXT",
    28: "AAAA", 33: "SRV", 35: "NAPTR", 43: "DS", 48: "DNSKEY", 64: "SVCB",
    65: "HTTPS", 255: "ANY",
}


# Pre-EDNS clients are limited to 512-byte UDP responses (RFC 1035).
_DEFAULT_UDP_SIZE = 512


def _question_end(query: bytes) -> int | None:
    """Offset of the byte just past the question section, or None if it can't be
    parsed. Used to trim a truncated response down to header + question."""
    qd = struct.unpack_from("!H", query, 4)[0]
    pos = 12
    for _ in range(qd):
        # QNAME: a sequence of length-prefixed labels ending in a zero length.
        # Compression pointers are illegal in a question, so reject them.
        while True:
            if pos >= len(query):
                return None
            length = query[pos]
            if length & 0xC0:
                return None
            pos += 1
            if length == 0:
                break
            pos += length
        pos += 4  # QTYPE + QCLASS
        if pos > len(query):
            return None
    return pos


def udp_payload_limit(query: bytes) -> int:
    """Largest UDP response the client will accept: its advertised EDNS0 payload
    size, or 512 when there is no OPT record. Best-effort; returns 512 on any
    trouble so we never over-promise."""
    try:
        ar = struct.unpack_from("!H", query, 10)[0]
        if ar == 0:
            return _DEFAULT_UDP_SIZE
        pos = _question_end(query)
        if pos is None:
            return _DEFAULT_UDP_SIZE
        # Walk the additional section looking for the OPT record (TYPE 41); its
        # CLASS field carries the requestor's UDP payload size.
        for _ in range(ar):
            if pos >= len(query):
                return _DEFAULT_UDP_SIZE
            if query[pos] & 0xC0 == 0xC0:        # compressed name
                pos += 2
            else:
                while pos < len(query) and query[pos] != 0:
                    pos += query[pos] + 1
                pos += 1                          # the zero length octet
            rtype, rclass = struct.unpack_from("!HH", query, pos)
            if rtype == 41:                       # OPT
                return max(_DEFAULT_UDP_SIZE, rclass)
            rdlen = struct.unpack_from("!H", query, pos + 8)[0]
            pos += 10 + rdlen                     # type+class+ttl+rdlen+rdata
        return _DEFAULT_UDP_SIZE
    except Exception:  # noqa: BLE001 - never raise on the resolve path
        return _DEFAULT_UDP_SIZE


def truncated_response(query: bytes) -> bytes | None:
    """Build a minimal TC=1 (truncated) response echoing the query's question, so
    an over-large answer makes the client retry over TCP instead of relying on IP
    fragmentation (which the DPI middleboxes we target routinely drop). Returns
    None if the query can't be parsed, so the caller can fall back."""
    end = None
    try:
        if len(query) >= 12:
            end = _question_end(query)
    except Exception:  # noqa: BLE001
        end = None
    if end is None:
        return None
    msg = bytearray(query[:end])
    msg[2] |= 0x80   # QR  = 1 (response)
    msg[2] |= 0x02   # TC  = 1 (truncated)
    msg[3] &= 0xF0   # clear RCODE (low nibble) -> NOERROR
    struct.pack_into("!HHH", msg, 6, 0, 0, 0)  # ANCOUNT = NSCOUNT = ARCOUNT = 0
    return bytes(msg)


def describe_query(query: bytes) -> str:
    """Return e.g. ``"example.com A"`` from a raw DNS query, or a best-effort
    fallback string if the message can't be parsed."""
    try:
        # Skip the 12-byte header, then read the QNAME labels.
        i = 12
        labels = []
        while True:
            length = query[i]
            i += 1
            if length == 0:
                break
            # On-wire labels are ASCII (IDNs arrive as xn-- punycode already).
            labels.append(query[i:i + length].decode("ascii", "replace"))
            i += length
        name = ".".join(labels) if labels else "."
        (qtype,) = struct.unpack_from("!H", query, i)
        return f"{name} {_QTYPES.get(qtype, f'TYPE{qtype}')}"
    except Exception:  # noqa: BLE001 - logging must never fail
        return f"<unpar? {len(query)}B>"

File: src/dnsutil.rs
//! Tiny helper to make DNS queries human-readable for logging. Never panics:
//! any malformed input yields a best-effort fallback string.

/// Common QTYPE numbers -> names (anything else is shown as `TYPE<n>`).
fn qtype_name(t: u16) -> String {
    match t {
        1 => "A".into(),
        2 => "NS".into(),
        5 => "CNAME".into(),
        6 => "SOA".into(),
        12 => "PTR".into(),
        15 => "MX".into(),
        16 => "TXT".into(),
        28 => "AAAA".into(),
        33 => "SRV".into(),
        35 => "NAPTR".into(),
        43 => "DS".into(),
        48 => "DNSKEY".into(),
        64 => "SVCB".into(),
        65 => "HTTPS".into(),
        255 => "ANY".into(),
        other => format!("TYPE{other}"),
    }
}

/// e.g. `"example.com A"` from a raw DNS query, or a fallback if unparseable.
pub fn describe_query(query: &[u8]) -> String {
    describe_inner(query).unwrap_or_else(|| format!("<unpar? {}B>", query.len()))
}

fn describe_inner(query: &[u8]) -> Option<String> {
    // Skip the 12-byte header, then read the QNAME labels.
    let mut i = 12usize;
    let mut labels: Vec<String> = Vec::new();
    loop {
        let len = *query.get(i)? as usize;
        i += 1;
        if len == 0 {
            break;
        }
        let label = query.get(i..i + len)?;
        // On-wire labels are ASCII (IDNs arrive as xn-- punycode already).
        labels.push(String::from_utf8_lossy(label).into_owned());
        i += len;
    }
    let name = if labels.is_empty() {
        ".".to_string()
    } else {
        labels.join(".")
    };
    let qtype = u16::from_be_bytes([*query.get(i)?, *query.get(i + 1)?]);
    Some(format!("{name} {}", qtype_name(qtype)))
}

#[cfg(test)]
mod tests {
    use super::*;

    // Header (id=0, flags, qd=1) + 7"example"3"com"0 + QTYPE A + QCLASS IN
    const EXAMPLE_A: &[u8] = b"\x00\x00\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\
\x07example\x03com\x00\x00\x01\x00\x01";

    #[test]
    fn parses_name_and_type() {
        assert_eq!(describe_query(EXAMPLE_A), "example.com A");
    }

    #[test]
    fn known_and_unknown_qtypes() {
        let mut aaaa = b"\x00\x00\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\x03abc\x00".to_vec();
        aaaa.extend_from_slice(&28u16.to_be_bytes()); // AAAA
        aaaa.extend_from_slice(&1u16.to_be_bytes());
        assert_eq!(describe_query(&aaaa), "abc AAAA");

        let mut weird = b"\x00\x00\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\x03abc\x00".to_vec();
        weird.extend_from_slice(&999u16.to_be_bytes()); // not in the table
        weird.extend_from_slice(&1u16.to_be_bytes());
        assert_eq!(describe_query(&weird), "abc TYPE999");
    }

    #[test]
    fn never_panics_on_garbage() {
        assert!(describe_query(&[]).starts_with("<unpar?"));
        assert!(describe_query(&[0u8; 5]).starts_with("<unpar?"));
        // Truncated label length running past the buffer.
        let q = b"\x00\x00\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\xff";
        assert!(describe_query(q).starts_with("<unpar?"));
    }
}
